internal/app/machined/main.go

```diff
@@ -80,217 +79,90 @@ func sync() {
-	if err = startup.RandSeed(); err != nil {
-		panic(err)
-	}
-
-	// Set the PATH env var.
-	if err = os.Setenv("PATH", constants.PATH); err != nil {
-		panic(errors.New("error setting PATH"))
+	p, err := platform.CurrentPlatform()
+	if err != nil {
+		reboot(err)
 	}
 
-	immediateReboot := false
+	r := runtime.NewRuntime(p, nil, runtime.Initialize)
 
-	// Boot the machine.
-	seq := sequencer.New(sequencer.V1Alpha1)
+	s := &v1alpha1runtime.Sequencer{}
 
-	if err := seq.Boot(); err != nil {
-		if errors.Is(err, runtime.ErrReboot) {
-			immediateReboot = true
-		} else {
-			log.Println(err)
-			panic(fmt.Errorf("boot failed: %w", err))
-		}
+	controller := &runtime.Controller{
+		Runtime:   r,
+		Sequencer: s,
 	}
 
-	// Wait for an event.
+	if p.Mode() != runtime.Container {
+		go func() {
+			if e := acpi.StartACPIListener(); e != nil {
+				log.Printf("WARNING: ACPI events will be ignored: %+v", err)
 
-	flag := unix.LINUX_REBOOT_CMD_RESTART
-
-	runShutdownOnImmediateReboot := true
+				return
+			}
 
-	for !immediateReboot {
-		switch e := <-init.Channel(); e.Type {
-		case event.Shutdown:
-			flag = unix.LINUX_REBOOT_CMD_POWER_OFF
-			fallthrough
-		case event.Reboot:
-			immediateReboot = true
-		case event.Upgrade:
-			var (
-				req *machineapi.UpgradeRequest
-				ok  bool
-			)
+			log.Printf("shutdown via ACPI received")
 
-			if req, ok = e.Data.(*machineapi.UpgradeRequest); !ok {
-				log.Println("cannot perform upgrade, unexpected data type")
-				continue
+			if e := controller.Run(runtime.Shutdown, nil); e != nil {
+				reboot(e)
 			}
+		}()
+	} else {
+		termCh := make(chan os.Signal, 1)
+		signal.Notify(termCh, syscall.SIGTERM)
 
-			if err := seq.Upgrade(req); err != nil {
-				log.Println(err)
-				panic(fmt.Errorf("upgrade failed: %w", err))
-			}
+		go func() {
+			<-termCh
+			signal.Stop(termCh)
 
-			immediateReboot = true
-			runShutdownOnImmediateReboot = false
-		case event.Reset:
-			var (
-				req *machineapi.ResetRequest
-				ok  bool
-			)
-
-			if req, ok = e.Data.(*machineapi.ResetRequest); !ok {
-				log.Println("cannot perform reset, unexpected data type")
-				continue
-			}
+			log.Printf("shutdown via SIGTERM received")
 
-			if err := seq.Reset(req); err != nil {
-				log.Println(err)
-				panic(fmt.Errorf("reset failed: %w", err))
+			if e := controller.Run(runtime.Shutdown, nil); e != nil {
+				reboot(e)
 			}
+		}()
+	}
 
-			if !req.GetReboot() {
-				flag = unix.LINUX_REBOOT_CMD_POWER_OFF
-			}
+	if err = controller.Run(runtime.Initialize, nil); err != nil {
+		reboot(err)
+	}
 
-			immediateReboot = true
+	go func() {
+		server := &v1alpha1server.Server{
+			Controller: controller,
 		}
-	}
 
-	if runShutdownOnImmediateReboot {
-		if err := seq.Shutdown(); err != nil {
-			log.Println(err)
-			panic(fmt.Errorf("shutdown failed: %w", err))
+		if e := factory.ListenAndServe(server, factory.Network("unix"), factory.SocketPath(universe.MachineSocketPath)); e != nil {
+			reboot(e)
```

Comment:
it doesn't seem to be super safe that we call this from goroutine, so that we don't give main goroutine any chance to clean up stuff? no technical reason for it right now, but basically it's not synced, so two concurrent `reboot` calls might be issued.

imho `main` should always exit with something like `reboot`